python/ray/actor.py

```diff
@@ -745,11 +717,21 @@ def _deserialization_helper(self, state, ray_forking):
         worker.check_connected()
 
         if state["ray_forking"]:
-            actor_handle_id = compute_actor_handle_id(
-                ray.ObjectID(state["previous_actor_handle_id"]),
-                state["actor_forks"])
+            actor_handle_id = ray.ObjectID(state["actor_handle_id"])
         else:
-            actor_handle_id = None
+            # Right now, if the actor handle has been pickled, we create a
+            # temporary actor handle id for invocations.
+            # TODO(pcm): This still leads to a lot of actor handles being
+            # created, there should be a better way to handle pickled
+            # actor handles.
+            # TODO(swang): Accessing the worker's current task ID is not
+            # thread-safe.
+            # TODO(swang): Unpickling the same actor handle twice in the same
+            # task will break the application, and unpickling it twice in the
+            # same actor is likely a performance bug. We should consider
+            # logging a warning in these cases.
+            actor_handle_id = compute_actor_handle_id_non_forked(
+                ray.ObjectID(state["actor_handle_id"]), worker.current_task_id)
```

Comment:
So you are assigning a permanent actor handle ID to an actor handle when it is unpickled?

I think the reason that I didn't do that initially was the following example,

```python
a = Actor.remote()

@ray.remote
def f():
    a.method.remote()

f.remote()
f.remote()
```

In this case, `a` is unpickled on the workers. If the two `f` tasks are executed on the same worker, the second task will depend on the first task, when really there should be no dependence between the two.

Does that make sense? This is probably a bit of an edge case.

Reply:
Ah, I see, I didn't think about this case. Is there a use case for it? I lean towards disallowing this because it makes it pretty hard to track actor handles and it seems like it could easily lead to a hard-to-explain performance bug.

Reply:
@robertnishihara I don't understand why the second `f` depends on the first `f`. Could you please explain that? thanks.

Reply:
Well, the second `a.method.remote()` ideally woudn't depend on the first call to `a.method.remote()`, but in this PR, the same actor handle will be used for both and so the second one will have to wait for the first one to execute because methods submitted from a single actor handle are executed in the order that they are submitted. Does that make sense?

Reply:
Do you mean if the 2 `f` tasks happen to be executed on the same worker, `a` will only be unpickled once (that is,   the second task will use the same actor handle from the first task)? 
I'm under the impression that unpickling `a` for the 2 tasks should be independent.

Reply:
Yes, if the 2 `f` tasks happen to be executed on the same worker, then `a` will only be unpickled once (and so the second task will use the same actor handle from the first task).

The remote function `f` is shipped to each worker only once and unpickled only once. And the handle `a` is unpickled as part of unpickling `f`.

Reply:
I see, thanks!